scripts/addr2line.py

```diff
@@ -54,7 +54,9 @@ class Addr2Line:
     dummy_pattern = re.compile(
         r"(.*0x0000000000000000: \?\? \?\?:0\n)"  # addr2line pattern
         r"|"
-        r"(.*0x0: \?\? at .*\n)"  # llvm-addr2line pattern
+        r"(.*0x0: \?\? at .*\n)"  # llvm-addr2line patterns
+        r"|"
+        r"(\?\? at \?\?:0\n)"
```

Comment:
If we are no longer using 0x0 as the sentinel, there's no point leaving this part of the pattern in the code, is there?

```suggestion
        r"(\?\? at \?\?:0\n)"
```

Reply:
I didn't check what's the output of older llvm-addr2line versions.
In the binutils'es `addr2line` the output for 0x0, "\n" and ",\n" is the same.

So, I assumed that the same probably holds for older versions of `llvm-add2line` and for them the pattern is as at the line 57 above.

I didn't want to break things. ;)

Reply:
Hmmm...
For `llvm-addr2line-10` the "coma trick" doesn't work.
Here is the output of `llvm-addr2line-10`:

```
$ echo 0x0 | llvm-addr2line-10 -Cfpia -e /usr/bin/bash
0x0: ?? at ??:0
$ echo "," | llvm-addr2line-10 -Cfpia -e /usr/bin/bash
,
$ echo "" | llvm-addr2line-10 -Cfpia -e /usr/bin/bash

$
```

Reply:
llvm-10 comes with U20.04

Reply:
U22.04 comes with llvm-14:

```
$ echo 0x0 | llvm-addr2line-14 -e /usr/bin/bash
??:0
$ echo "," | llvm-addr2line-14 -e /usr/bin/bash
,
$ echo "" | llvm-addr2line-14 -e /usr/bin/bash

$
```

Reply:
Yes, LLVM <= 17 just echoes invalid lines. If you want to support older versions of LLVM, I guess you can just an extra `r"(,\n)"` pattern to `dummy_pattern`.

It seems that LLVM >= 18 adjusted its behaviour for invalid lines to match binutils's format.

Reply:
Validation:
```
~ $ for ver in (seq 12 19);
      nix shell nixpkgs#llvmPackages_$ver.llvm --command bash -c 'llvm-addr2line --version | grep version;
      llvm-addr2line -Cipfe /bin/sh <<< ,';
      end
  LLVM version 12.0.1
,
  LLVM version 13.0.1
,
  LLVM version 14.0.6
,
  LLVM version 15.0.7
,
  LLVM version 16.0.6
,
  LLVM version 17.0.6
,
  LLVM version 18.1.8
?? at ??:0
  LLVM version 19.1.6
?? at ??:0
```

Reply:
Conditioning on `llvm-addr2line` version is going to uglify the code:
* The most platform-independent way to do it is to go via `llvm-addr2line --version` and then parse the output.
* But its output seems to be not uniform:

```
$ llvm-addr2line-14 --version
llvm-addr2line
Ubuntu LLVM version 14.0.0
  
  Optimized build.
  Default target: x86_64-pc-linux-gnu
  Host CPU: znver3
$

$ llvm-addr2line-18 --version
llvm-addr2line
Ubuntu LLVM version 18.1.3
  Optimized build.
$

$ llvm-addr2line-10 --version
LLVM (http://llvm.org/):
  LLVM version 10.0.0
  
  Optimized build.
  Default target: x86_64-pc-linux-gnu
  Host CPU: cascadelake
$
```

It looks like the version always comes in a second line after a word "version".
But assuming this feels too hackish.
But if that's what you want me to do - I'll do it ;)

Reply:
Wouldn't this work?:

```suggestion
        r"(,\n)"  # llvm-addr2line patterns
        r"|"
        r"(\?\? at \?\?:0\n)"
```

I think this pattern should work as we want with both the "old LLVM style" and the "new LLVM style".

And I share your pain that it's all so hacky. But it seems that the only non-hacky alternative is to entirely replace addr2line with something else. (And that might end up more painful than the hacks).

Reply:
I have had to modify/fix this many times, and have added the `-v` argument to help debug these cases, so you can see when the parsing stalls out due to the failure to match the dummy line. I have a similar fix to this one in a branch somewhere because I ran into the same issue (can't recall exactly why I did not upstream it).

I think we should just make the regex more accepting rather than try to match the dummy line in a addr2line version-dependent way. The chance of a false positive is low.

I'm also wondering if we should just pass in a random 64-bit address on the kernel size of the address space (MSB set). If we see this in the output we know it is sentinel: such an address will never appear in a valid userspace stack trace (and, practically never on the kernel side either given the size of 64-bit space).

Reply:
> I think this pattern should work as we want with both the "old LLVM style" and the "new LLVM style".

That's actually a (brilliantly simple) good idea!

Reply:
done

Reply:
Wait! You wanted to also remove the no-longer-needed patters.

Reply:
Now it's proper-done.

Reply:
@travisdowns your idea seems to be overly complicated compared to ","-line sentinel one.

Reply:
> @travisdowns your idea seems to be overly complicated compared to ","-line sentinel one.

Sure, whatever is simplest: I just really want to avoid using tool-version specific regex (e.g., parsing `--version`). If you've found a trick all current binutils and llvm addr2line versions that's awesome!